sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+- Low-level APIs efficiently mapped to Gen architecture, such as block reads.
+
+One of the future directions is enabling this extension for other architectures,
+such as x86, with extracting and clearly marking generic and target-dependent
+API portions.
+
+## Explicit SIMD execution model
+
+Explicit SIMD execution model is a basically an equivalent of the base SYCL
+execution model with subgroup size restricted to 1. Which means in ESP
+it is guaranteed that there is a single work item in any workgroup, and it maps
+to a single hardware thread. All standard SYCL APIs continue to work,
+including `sycl::intel::sub_group` ones, which become either a no-op or
+trivial. E.g. a barrier becomes just a memory fence for a compiler, collectives
+just return the value in the single work-item. Another consequences of the unit
+subgroup size is guaranteed independent forward progress between work-items on
+some Gen architecture generations.
+
+## Explicit SIMD extension APIs
+
+Explicit SIMD APIs can be used only in code to be executed on Intel Gen
+architecture devices and the host device for now. Attempt to run such code on
+other devices will result in error. 
+
+Kernels and `SYCL_EXTERNAL` functions using ESP must be explicitly marked with
+the `EXPLICIT_SIMD` attribute macro. Subgroup size quiery within such
+functions will always return `1`.
+
+*Functor kernel*
+```cpp
+using AccTy = cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
+                                 cl::sycl::access::target::global_buffer>;
+class Functor {
+public:
+  Functor(int X, AccTy &Acc) : X(X), Acc(Acc) {}
+
+  EXPLICIT_SIMD void operator()() { Acc[0] += X; }
+
+private:
+  int X;
+  AccTy Acc;
+};
+```
+
+*Lambda kernel and function*
+```cpp
+EXPLICIT_SIMD SYCL_EXTERNAL
+void sycl_device_f(cl::sycl::global_ptr<int> ptr, sycl::intel::gpu::simd<float, 8> X) {
+  sycl::intel::gpu::flat_block_write(*ptr.get(), X);
+}
+...
+  Q.submit([&](cl::sycl::handler &Cgh) {
+    auto Acc1 = Buf1.get_access<cl::sycl::access::mode::read>(Cgh);
+    auto Acc2 = Buf2.get_access<cl::sycl::access::mode::read_write>(Cgh);
+
+    Cgh.single_task<class KernelID>([=] () EXPLICIT_SIMD {
+      sycl::intel::gpu::simd<float, 8> Val = sycl::intel::gpu::flat_block_read(Acc1.get_pointer());
+      sycl_device_f(Acc2, Val);
+    });
+  });
+}
+```
+
+## Implementation restrictions
+Current ESP implementation does not support using certain standard SYCL features
+inside explicit SIMD kernels and functions. Most of them will be eventually
+dropped. What's not supported today:
+- Mixing `EXPLICIT_SIMD` kernels with SYCL kernels in a single source
+- Local accessors. Local memory is allocated and accessed via explicit
+device-side API
+- 2D and 3D accessors
+- Constant accessors
+- `sycl::accessor::get_pointer()`. All memory accesses through an accessor are
+done via explicit APIs; e.g. `sycl::intel::gpu::block_store(acc, offset)`
+- Few others (to be documented)
```

Comment:
No. This will be added incrementally.

Reply:
Are they unknown or why would we not document all of them? If all known ones are documented that's fine.

Reply:
Some potential ones are known - need to be checked before specifying. I admit there can be some unknowns. 
This is WIP rather than a complete formal specification, will be improved/extended over time.